ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/objectstore/hornetq/HornetqJournalStore.java

```diff
@@ -182,15 +182,15 @@ public boolean write_committed(Uid uid, String typeName, OutputObjectState txDat
             long id = getId(uid, typeName);
             byte[] data = outputBuffer.buffer();
 
-            // yup, there is a race condition here.
-            if(getContentForType(typeName).containsKey(uid)) {
+            RecordInfo record = new RecordInfo(id, RECORD_TYPE, data, false, (short)0);
+            RecordInfo previousRecord = getContentForType(typeName).putIfAbsent(uid, record);
```

Comment:
We would need to remove this if the append failed, happy - can be added to the existing exception handler (line 195)